[TypePromotion] Do not zero-extend getelementptr indexes since signed

antoniofrighetto opened the pull request:
A miscompilation issue has been addressed with improved handling.

Fixes: https://github.com/llvm/llvm-project/issues/133928.

Alive2 for signext_bitcast_phi_select test: https://alive2.llvm.org/ce/z/gcMNvS.

antoniofrighetto commented:
This should match the original intent of the function, as per comment. We don't seem to bother with sign-extending here, but can change this to sext instead of bailing out, if needed.

antoniofrighetto commented:
@davemgreen Is there any particular reason why we do not run isTriviallyDead here in TypePromotion, as happens in other parts in CodeGen? We convert an unused trunc to a used zext in the snippet in https://github.com/llvm/llvm-project/issues/133928. This would impact 22 tests primarily among AArch64 and RISCV, so I'd like this to be confirmed before adding it.

davemgreen commented:
Hi - I would guess that we usually assume that trivial instructions have already been removed in anything that wasn't a hand-written (or reduced) test case, and that the snippet for #133928 would just need to be larger if it had to include extra uses for the trivial instructions - the original problem would remain just in a different form. I don't think it would be an issue if you wanted to add it, but I doubt it would be super useful in the entire pass pipeline.

I think in practice all gep indices will be canonicalized to ptrsize so this is probably OK. Can you update the `// Ignore GEPs because ...` comment before the gep, and add a testcase from #133928 with a comment explaining what is going on?

antoniofrighetto commented:
> Hi - I would guess that we usually assume that trivial instructions have already been removed in anything that wasn't a hand-written (or reduced) test case, and that the snippet for https://github.com/llvm/llvm-project/issues/133928 would just need to be larger if it had to include extra uses for the trivial instructions - the original problem would remain just in a different form. I don't think it would be an issue if you wanted to add it, but I doubt it would be super useful in the entire pass pipeline.

That makes sense, although perhaps some earlier pruning in TypePromotion could alleviate a bit of work in CGP, which is scheduled to run next and recursively deletes trivially dead instructions.

> I think in practice all gep indices will be canonicalized to ptrsize so this is probably OK. Can you update the // Ignore GEPs because ... comment before the gep, and add a testcase from https://github.com/llvm/llvm-project/issues/133928 with a comment explaining what is going on?

Updated, thanks!

davemgreen reviewed (approved):
Thanks. LGTM

antoniofrighetto committed:
[TypePromotion] Do not zero-extend getelementptr indexes since signed

A miscompilation issue has been addressed with improved handling.

Fixes: https://github.com/llvm/llvm-project/issues/133928.

Alive2: https://alive2.llvm.org/ce/z/gcMNvS.
